Add an endpoints reconciler component for external workloads

mateiidavid opened the pull request:
We introduced an endpoints controller that will be responsible for managing EndpointSlices for services that select external workloads. We introduce as a follow-up the reconciler component of the controller that will be responsible for doing the writes and diffing.

We also add callbacks for endpointslices. We want to ensure that changes to slices will be reconciled to avoid a user accidentally changing configuration. This will not result in a loop since we take care to check the newly observed resourceVersion against a record that we keep in memory (and that we update with every write, i.e. endpointSliceTracker).

There are a couple of key differences between what we do and what the upstream controller does.

1. Ports:

The upstream controller considers undocumented ports as valid as long as a service selects it. For example:

```yaml
kind: Service
spec:
  ports:
  - targetPort: 8080
    port: 80
```

would work even if the pod does not expose port 80. We require users to explicitly document their ports, and as a result, will skip processing any workloads that do not have a matching port.

2. More on ports... sort of:

Services can hold _lots_ of ports. As a result, the upstream controller is written in such a way that allows for a lot of ports to be shared among all of the endpointslices. Max capacity per slice is 100 ports (however a service may have more).

We work with this limitation instead and assume no service will have that many ports; if it does, it will error out.

3. And some more ports:

Changing a service's ports at runtime should result in:
- A create (we create a new slice with new ports)
- A delete (we delete the old slice with the old ports)

The upstream controller has some logic to ensure a create and a delete are turned into an update. We take a shortcut and just re-create the services for simplicity.

Testing is going to be similar to https://github.com/linkerd/linkerd2/pull/11940#issue-2082354571. We'll now be able to see endpointslices being written though. This is still WIP since we need to add some module tests.

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

adleong commented on controller/api/destination/external-workload/endpoints_reconciler.go:
```diff
@@ -0,0 +1,624 @@
+	// they segment slices by port sets.
+	// see: https://github.com/kubernetes/kubernetes/issues/99382
+	//
+	// We do not segment slices based on ports since we do not support a service
+	// with more than 100 ports.
```
Do we need to segment sliced based on port for correctness if different EW have named ports which resolve to different ports?  e.g. consider a service which defines a port named `http`.  The service selects two EW, one with a container port named `http` with port value 8000 and the other EW has a container port named `http` with port value 8001.  Would these two EW need to go in different slices because their endpoints use different ports?

mateiidavid replied:
Yup, good spot. Added in.

adleong commented on controller/api/destination/external-workload/endpoints_reconciler.go:
```diff
@@ -0,0 +1,624 @@
+
+	if len(desiredEndpointPorts) > maxEndpointPortsQuota {
+		// TODO: we should ensure this is bubbled up and visible to users, not
+		// sure the error msg makes sense.
```
TODO: may want a counter for this too

mateiidavid replied:
Port quota's no longer a thing since we now segment by ports.

zaharidichev reviewed (commented):
This is all coming together pretty well. On the testing strategy I think we need to focus on: 

Module tests for all major functions: 

- reconcileIPv4Endpoints
- reconcileEndpoints
- all the other util functions

More comprehensive correctness tests: 
- test that logic to prevent cache stalenss actually prevents it
- test situations where we assume the controller has missed a number of updates due to being down and then fires up from cold start
- test that controller does not modify any endpoint slices that are managed by other controllers


For the cold start, here is an approach that might work.

- Have two instances of the controller
- Have one of the controller process all events to arrive at a certain state of the API
- Have the second controller start at a later stage only seeing the final state of the API server
- Ensure that the state of the APIs is identical for both controller. 


Source of inspiration (upstream): 

Reconciler tests: https://github.com/robscott/kubernetes/blob/c75787bb778d45db1b2ee89cc15bf19306c02eb2/pkg/controller/endpointslice/reconciler_test.go
Controller tests: https://github.com/robscott/kubernetes/blob/c75787bb778d45db1b2ee89cc15bf19306c02eb2/pkg/controller/endpointslice/endpointslice_controller_test.go
Slice tracker test to verify cache staleness is caught: https://github.com/robscott/kubernetes/blob/c75787bb778d45db1b2ee89cc15bf19306c02eb2/pkg/controller/endpointslice/endpointslice_tracker_test.go#L26

adleong commented on controller/api/destination/external-workload/endpoints_reconciler.go:
```diff
@@ -0,0 +1,656 @@
+
+	// Add slices that we have already marked for deletion due to various
+	// reasons
+	totalToDelete = append(totalToDelete, toDelete...)
```
what's the purpose of having a totalToDelete and a toDelete as separate slices? why not just start with a slicesToDelete from the beginning and use it throughout (like in the upstream)?

adleong commented on controller/api/destination/external-workload/endpoints_reconciler.go:
```diff
@@ -0,0 +1,656 @@
+	// by ports. Besides risking to hit a port quota (of 100 ports per
+	// endpoint), we can have cases where a service selects a named port that
+	// maps to a different value on multiple workloads.
+	currentSlicesByPorts := map[epsliceutil.PortMapKey][]*discoveryv1.EndpointSlice{}
```
oh... is desiredEndpointPorts our equivalent of desiredMetaByPortMap??

adleong reviewed (commented):
I'm really struggling to review this because going through it side-by-side with https://github.com/kubernetes/endpointslice/blob/master/reconciler.go, it's tough to see exactly where things line up vs where they're different.  This is because we're using subtly different names for variables and funcitons so it can be hard to see what corresponds to what.

I would expect that these two implementations should be very very similar, with only pod-specific logic swapped out for externalworkload-specific logic.  If we could use the same variable and function names whenever possible, it would make it a lot easier to see exactly where we do and don't differ from the upstream.  I might even go as far as starting with a wholesale copy-paste of the upstream and then making explicit targeted changes where appropriate.

zaharidichev commented:
@mateiidavid @adleong

I managed to get the controller into in invalid state while doing some functional testing. Here is the repro: 

1. Apply the following diff so you get it all wired: 


```diff
diff --git a/controller/cmd/destination/main.go b/controller/cmd/destination/main.go
index 845c9e03c..17fd6cc03 100644
--- a/controller/cmd/destination/main.go
+++ b/controller/cmd/destination/main.go
@@ -9,6 +9,7 @@ import (
        "syscall"

        "github.com/linkerd/linkerd2/controller/api/destination"
+       externalworkload "github.com/linkerd/linkerd2/controller/api/destination/external-workload"
        "github.com/linkerd/linkerd2/controller/api/destination/watcher"
        "github.com/linkerd/linkerd2/controller/k8s"
        "github.com/linkerd/linkerd2/pkg/admin"
@@ -107,7 +108,7 @@ func Main(args []string) {
                        *kubeConfigPath,
                        true,
                        "local",
-                       k8s.Endpoint, k8s.ES, k8s.Pod, k8s.Svc, k8s.SP, k8s.Job, k8s.Srv,
+                       k8s.Endpoint, k8s.ES, k8s.Pod, k8s.Svc, k8s.SP, k8s.Job, k8s.Srv, k8s.ExtWorkload,
                )
        } else {
                k8sAPI, err = k8s.InitializeAPI(
@@ -115,7 +116,7 @@ func Main(args []string) {
                        *kubeConfigPath,
                        true,
                        "local",
-                       k8s.Endpoint, k8s.Pod, k8s.Svc, k8s.SP, k8s.Job, k8s.Srv,
+                       k8s.Endpoint, k8s.Pod, k8s.Svc, k8s.SP, k8s.Job, k8s.Srv, k8s.ExtWorkload,
                )
        }
        if err != nil {
@@ -159,6 +160,12 @@ func Main(args []string) {
        metadataAPI.Sync(nil)
        clusterStore.Sync(nil)

+       ec, err := externalworkload.NewEndpointsController(k8sAPI, "my-hostname", *controllerNamespace, done)
+       if err != nil {
+               log.Fatalf("Failed to initialize EndpointsController: %s", err)
+       }
+       ec.Start()
+
        go func() {
                log.Infof("starting gRPC server on %s", *addr)
                if err := server.Serve(lis); err != nil {

```

2. Apply the following resources `kubectl apply -f ext.yml`

ext.yml: 

```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-svc
spec:
  selector:
    app: external
  ports:
    - name: http
      protocol: TCP
      port: 80
---
 apiVersion: workload.linkerd.io/v1alpha1
 kind: ExternalWorkload
 metadata:
  name: my-ext-workload
  labels:
    app: external
 spec:
  meshTls:
    serverName: "identity.local"
    identity: "spiffe://identity.com"
  workloadIPs:
  - ip: "192.0.2.1"
  ports:
  - port: 80
 status:
  conditions:
   - type: Ready
     status: "True"
```

3. Observe endpoints: 
```bash
$ kubectl get endpointslices
NAME                           ADDRESSTYPE   PORTS     ENDPOINTS      AGE
kubernetes                     IPv4          6443      192.168.96.2   33m
my-svc-nddbf                   IPv4          <unset>   <unset>        4s
linkerd-external-my-svcbtcm7   IPv4          80        192.0.2.1      4s

```

4. Now do: 
```bash
$ kubectl delete -f ../ext.yml
service "my-svc" deleted
externalworkload.workload.linkerd.io "my-ext-workload" deleted
```

5. State of endpoints slices: 
```bash
$ kubectl get endpointslices
NAME         ADDRESSTYPE   PORTS   ENDPOINTS      AGE
kubernetes   IPv4          6443    192.168.96.2   35m
```

6. Now try and recreate: 
```bash
$ kubectl apply -f ../ext.yml
service/my-svc created
externalworkload.workload.linkerd.io/my-ext-workload created
```

7. State of endpoints slices: 
```bash
$ kubectl get endpointslices
NAME           ADDRESSTYPE   PORTS     ENDPOINTS      AGE
kubernetes     IPv4          6443      192.168.96.2   36m
my-svc-hzc5f   IPv4          <unset>   <unset>        26s


$ kubectl get endpointslices my-svc-hzc5f -oyaml
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints: null
kind: EndpointSlice
metadata:
  annotations:
    endpoints.kubernetes.io/last-change-trigger-time: "2024-01-21T15:26:01Z"
  creationTimestamp: "2024-01-21T15:26:01Z"
  generateName: my-svc-
  generation: 1
  labels:
    endpointslice.kubernetes.io/managed-by: endpointslice-controller.k8s.io
    kubernetes.io/service-name: my-svc
  name: my-svc-hzc5f
  namespace: default
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Service
    name: my-svc
    uid: 3e7658f1-de7f-48de-a1e3-3e66f53d5f4f
  resourceVersion: "3463"
  uid: 180d4e30-79a3-4029-8975-d43e9cf4ace8
ports: null
```

8. Logs: 

```bash
kubectl logs deploy/linkerd-destination  -c destination  -n linkerd  -f
time="2024-01-21T15:19:43Z" level=info msg="running version dev-983fc55a-zaharidichev"
time="2024-01-21T15:19:43Z" level=info msg="starting admin server on :9996"
time="2024-01-21T15:19:43Z" level=info msg="Using default opaque ports: map[25:{} 587:{} 3306:{} 4444:{} 5432:{} 6379:{} 9300:{} 11211:{}]"
time="2024-01-21T15:19:43Z" level=warning msg="failed to register Prometheus gauge Desc{fqName: \"job_cache_size\", help: \"Number of items in the client-go job cache\", constLabels: {cluster=\"local\"}, variableLabels: {}}: duplicate metrics collector registration attempted"
time="2024-01-21T15:19:43Z" level=info msg="waiting for caches to sync"
time="2024-01-21T15:19:43Z" level=info msg="caches synced"
time="2024-01-21T15:19:43Z" level=info msg="waiting for caches to sync"
time="2024-01-21T15:19:43Z" level=info msg="caches synced"
time="2024-01-21T15:19:43Z" level=info msg="waiting for caches to sync"
time="2024-01-21T15:19:43Z" level=info msg="caches synced"
time="2024-01-21T15:19:43Z" level=info msg="starting gRPC server on :8086"
time="2024-01-21T15:19:43Z" level=info msg="attempting to acquire leader lease linkerd/linkerd-destination-endpoint-write..."
time="2024-01-21T15:19:43Z" level=info msg="successfully acquired lease linkerd/linkerd-destination-endpoint-write"
time="2024-01-21T15:19:43Z" level=info msg="my-hostname acquired lease" component=external-endpoints-controller
```

9. Now, restart the destination controller and check state of slices again: 

```bash
$ kubectl  rollout restart deploy/linkerd-destination -n linkerd
deployment.apps/linkerd-destination restarted

$ kubectl get endpointslices
NAME                           ADDRESSTYPE   PORTS     ENDPOINTS      AGE
kubernetes                     IPv4          6443      192.168.96.2   39m
my-svc-hzc5f                   IPv4          <unset>   <unset>        2m59s
linkerd-external-my-svc2fw56   IPv4          80        192.0.2.1      2s
```

**That obviously looks invalid.**

zaharidichev reviewed (commented):
Good work. Writing controllers with this framework is complicated. I have spotted some bugs that lead us to an invalid state. Additionally, I agree with @adleong that we might want to source more inspiration from the upstream implementation.  Also, functional testing is a good way to find bugs, understand the code better and get inspiration for unit tests.  Here are my suggestions at the moment: 

- go through the problems I have encountered and pointed out and try to understand why they happen  (I have added some pointers).
- try to see whether some of these problems occur because we diverge from upstream implementation at times.
- look out for methods on the EndpointSlicesTracker that they are using upstream, but we are not (e.g. `DeleteService`) and think about why we are not using them
- seems to me that a lot of the problems stem from the complicated state management of the endpoints tracker. Would it be useful to leave this bit out and have it as a follow-up PR like I suggested in the beginning? This might make things easier to think through and reason about.

Let's chat about all this tomorrow. And again, writing such controllers is hard. Lets try and source as much inspiration from the upstream implementation. The state of their code is the culmination of multiple bug-hunting expeditions and small improvements. We can try and minimise that by following their patterns more closely as @adleong recommends as well.

mateiidavid commented:
Thanks @zaharidichev and @adleong for reviews! Rather than address each individual comment one-by-one, I'm going to reply in a comment to get both of you up-to-date on the state of the PR.

Since we all think it's better to try and minimise how much we diverge from the upstream implementation, I've done another pass through the reconciler to ensure we're as close as possible to the source. This has lead me to re-think some of our assumptions and constraints. I won't list them all, but I'll mention the most important ones:
* Ports! One of our key differentiations was that a workload port _must_ be documented. However, this is only something we introduced outside of what the spec (and contract) of an external workload enforces. Since ports are optional, it makes sense to treat them exactly the same as a pod's ports are treated. We essentially do the same thing for policy discovery, so it made a lot of sense to me to simplify this part of our code and do exactly what the upstream does when a service uses a numeric targetPort.
* Optimisation #1: instead of a create and delete do an update. This is something the upstream does. I felt like it complicates the code. On second thought, it's an easy enough change to test so I've added it in. For example:
  * a service selecting a workload will create an endpoint slice. When (and if) the service's targetPort would change, the old behaviour would be to create a new slice and delete the old one.
  * we are now doing the same thing as upstream; we update the slice in place.
* More ports! Instead of failing when a port is not matching (e.g. a targetPort references a name that does not exist on a considered workload) we allow the endpointslice to set an empty ports list on the slice. This is the same as upstream.
* Optimisation #2: we pack unchanged slices more efficiently (i.e. same thing upstream does).

It's maybe easier to tell you what we _don't do_. We categorically do not create a placeholder slice. I don't think this would help at all with state tracking, and I don't think it fits our API. We should conditionally create slices when a service matches a workload. Since we do not need to provide a service discovery API to other components, I don't this really applies to us.

mateiidavid commented on controller/api/destination/external-workload/endpoints_reconciler.go:
```diff
@@ -0,0 +1,624 @@
+		// Loop through the endpointslices and figure out which endpointslice
+		// does not have an ownerRef set to the service. If a slice has been
+		// selected but does not point to the service, we delete it.
+		if !ownedBy(slice, svc) {
```
Nope, not really. This ensures that ownerReferences are set-up properly, but we still verify we only work with objects whose owner is this service in the prior processing step. e.g. we get workloads with the service's label selector. We also get endpoint slices only for our service and only slices managed by our controller. (but we can't verify their owner refs until we reach this point).

adleong commented on controller/api/destination/external-workload/controller_util.go:
```diff
@@ -0,0 +1,270 @@
+	// Service selections (i.e. services that select a workload through a label
+	// selector) may reference an old workload, a new workload, or both,
+	// depending on the workload's labels.
+	if labelsChanged && specChanged {
```
Having a hard time comparing this to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/endpointslice/util/controller_utils.go#L184 and convincing myself that they're equivalent.  Can they be made more similar to make this easier?  A nice bonus of this is we get to use "k8s.io/apimachinery/pkg/util/sets" instead of hand rolled sets.

zaharidichev replied:
will restructure

adleong commented on controller/api/destination/external-workload/controller_util.go:
```diff
@@ -0,0 +1,270 @@
+	if isReady(old) != isReady(updated) {
+		return true
+	}
+
```
I noticed that the upstream podEndpointsChanged returns that the spec has changed if the deletion timestamps differ and we don't. Is that intentional?

adleong commented on controller/api/destination/external-workload/endpoints_reconciler.go:
```diff
@@ -0,0 +1,704 @@
+
+	// If there are any slices whose ports no longer match what we want in our
+	// current reconciliation, delete them
+	for _, existingSlice := range existingSlices {
```
can't we use currentSlicesByPorts instead of existingSlices so that we don't have to recompute the portHashes again?  As in https://github.com/kubernetes/endpointslice/blob/master/reconciler.go#L225

zaharidichev commented on controller/api/destination/external-workload/controller_util.go:
```diff
@@ -0,0 +1,270 @@
+	if !ok {
+		return nil, fmt.Errorf("couldn't get ExternalWorkload from object %#v", o)
+	}
+
```
Good catch

zaharidichev commented on controller/api/destination/external-workload/endpoints_controller.go:
```diff
@@ -437,15 +437,15 @@ func (ec *EndpointsController) onAddExternalWorkload(obj interface{}) {
 	services := ec.getServicesToUpdateOnExternalWorkloadChange(old, cur)
 
-	for _, svc := range services {
+	for svc := range services {
```
This was resulting in a panic whenever these callbacks fire:


```
time="2024-01-24T08:12:27Z" level=info msg="waiting for caches to sync"
time="2024-01-24T08:12:27Z" level=info msg="caches synced"
time="2024-01-24T08:12:27Z" level=info msg="waiting for caches to sync"
time="2024-01-24T08:12:27Z" level=info msg="caches synced"
time="2024-01-24T08:12:27Z" level=info msg="waiting for caches to sync"
time="2024-01-24T08:12:27Z" level=info msg="caches synced"
time="2024-01-24T08:12:27Z" level=info msg="starting gRPC server on :8086"
panic: interface conversion: interface {} is sets.Empty, not string

goroutine 287 [running]:
github.com/linkerd/linkerd2/controller/api/destination/external-workload.(*EndpointsController).processQueue(0xc000274790)
        /linkerd-build/controller/api/destination/external-workload/endpoints_controller.go:235 +0x105
created by github.com/linkerd/linkerd2/controller/api/destination/external-workload.(*EndpointsController).Start in goroutine 1
        /linkerd-build/controller/api/destination/external-workload/endpoints_controller.go:218 +0x125
```

Again, we need to resort to functional testing quite a bit here

mateiidavid replied:
Nice, I managed to catch this separately when I fired up the controller. Fixed and confirmed it works.
